### PR TITLE
[coop] Transition all functions in object.h to MONO_RT_EXTERNAL_ONLY

### PR DESCRIPTION
**Pull request by jaykrell**
In general this is achieved by renaming mono_foo to mono_foo_internal,
and mono_foo calls mono_foo_internal, possibly adding a GC mode change
to mono_foo and removing from mono_foo_internal.

Going forward though the wrapping could diverge further. For example, all the _internal functions could only accept/return coop handles, whereas currently they handle a lot of raw pointers.

This gives us leeway to limit gc mode changes to embedders,
and more generally to separate the runtime from the embedding API.

### CODE REVIEW

File: mono/metadata/object.h [hunk 1]
```diff
@@ -230,7 +230,7 @@ mono_monitor_enter           (MonoObject *obj);
 MONO_API void
 mono_monitor_enter_v4        (MonoObject *obj, char *lock_taken);
 
-MONO_API unsigned int
+MONO_API MONO_RT_EXTERNAL_ONLY unsigned
```
Comment (luhenry) [inline mono/metadata/object.h]: Revert that to `unsigned int`

File: mono/metadata/object.h [hunk 2]
```diff
@@ -175,13 +175,13 @@ mono_string_from_utf16	    (/*const*/ mono_unichar2 *data);
 MONO_API MONO_RT_EXTERNAL_ONLY MonoString *
 mono_string_from_utf32	    (/*const*/ mono_unichar4 *data);
 
-MONO_API mono_bool
+MONO_API MONO_RT_EXTERNAL_ONLY mono_bool
 mono_string_equal           (MonoString *s1, MonoString *s2);
 
-MONO_API unsigned int
+MONO_API MONO_RT_EXTERNAL_ONLY unsigned
```
Comment (luhenry) [inline mono/metadata/object.h]: Revert that to `unsigned int`.

File: mono/metadata/object.h [hunk 3]
```diff
@@ -69,11 +69,11 @@ typedef void	    (*MonoMainThreadFunc)    (void* user_data);
...
 	} while (0)
 
-MONO_API mono_unichar2 *mono_string_chars  (MonoString *s);
-MONO_API int            mono_string_length (MonoString *s);
+MONO_API MONO_RT_EXTERNAL_ONLY mono_unichar2 *mono_string_chars  (MonoString *s);
+MONO_API MONO_RT_EXTERNAL_ONLY int            mono_string_length (MonoString *s);
 
 MONO_API MONO_RT_EXTERNAL_ONLY MonoObject *
-mono_object_new		    (MonoDomain *domain, MonoClass *klass);
+mono_object_new (MonoDomain *domain, MonoClass *klass);
```
Comment (luhenry) [inline mono/metadata/object.h]: Revert that.

File: mono/metadata/object.h (hunk 1 above)
Comment (jaykrell) [reply mono/metadata/object.h]: Same thing.

**Comment by jaykrell** [comment]
Something wrong with profiler tests.

**Comment by jaykrell** [comment]
Crash on this PR was apparently the hint to resolve csc crashes.

https://jenkins.mono-project.com/job/test-mono-pull-request-interpreter/11028/parsed_console/log.html

```Thread 20 (Thread 0x7f82c3fff700 (LWP 25840)):
#0  0x00007f8323c95b3a in waitpid () from /lib/x86_64-linux-gnu/libpthread.so.0
#1  0x00005588a228e9a6 in dump_native_stacktrace (ctx=ctx@entry=0x7f8319163ac0, signal=0x5588a2472b60 "SIGSEGV") at mini-posix.c:1092
#2  0x00005588a228eb20 in mono_dump_native_crash_info (signal=signal@entry=0x5588a2472b60 "SIGSEGV", ctx=ctx@entry=0x7f8319163ac0, info=info@entry=0x7f8319163bf0) at mini-posix.c:1115
#3  0x00005588a2225575 in mono_handle_native_crash (signal=signal@entry=0x5588a2472b60 "SIGSEGV", ctx=ctx@entry=0x7f8319163ac0, info=info@entry=0x7f8319163bf0) at mini-exceptions.c:3196
#4  0x00005588a21a7374 in mono_sigsegv_signal_handler_debug (_dummy=11, _info=0x7f8319163bf0, context=0x7f8319163ac0, debug_fault_addr=0xfcb36) at mini-runtime.c:3527
#5  <signal handler called>
#6  0x00007f8320072432 in System_Reflection_Metadata_BlobBuilder__ctor_int (this=..., capacity=1024) from /mnt/jenkins/workspace/test-mono-pull-request-interpreter/mcs/class/lib/build/System.Reflection.Metadata.dll.so
#7  0x00007f832039282f in Microsoft_Cci_PooledBlobBuilder__ctor_int (this=..., size=1024) from /mnt/jenkins/workspace/test-mono-pull-req```

**Comment by lambdageek** [comment]
Could you explain why `object.c` includes `external-only.c` - why can't we compile `external-only.c` as a regular object file?

**Comment by jaykrell** [comment]
It seems to break things. It makes sense that it breaks things on Windows. I thought Unix would work.

On Windows, the normal static linker, functions marked __declspec(dllexport) are only exported, if you otherwise get the linker to bring in anything from the .obj, and then it sees the directives.

.def files do not have this problem and I have recommended them repeatedly.
With .def  files, the linker must resolve the symbol and export it, not just export it if it is found.

I thought Unix linkers were more eager, so was surprised at the breakage, but anyway, you  have to do it this way for Windows.

Another option is to list the .o/.obj on the link command line in mini, instead of burying it in a lib.

There is also a  very new switch to the Windows linker /wholearchive that makes it more Unix-like.

(and I forgot to update the project files again)

**Comment by jaykrell** [comment]
Profiler test failure not reproing locally.

File: mono/metadata/object.c [hunk 4]
```diff
@@ -55,6 +55,9 @@
 #include <mono/utils/w32api.h>
 #include <mono/utils/unlocked.h>
 
+// Export do not work if compiled separately.
+#include "external-only.c"
```
Comment (luhenry) [inline mono/metadata/object.c]: Well, I guess you will have to revert some of the breaking changes then 😉

**Comment by jaykrell** [comment]
@monojenkins build Linux i386

**Comment by jaykrell** [comment]
@monojenkins build Linux x64

**Comment by jaykrell** [comment]
hm same failures with reduced change

File: mono/metadata/object.c [hunk 5]
```diff
@@ -2834,19 +2852,23 @@ mono_upgrade_remote_class (MonoDomain *domain, MonoObjectHandle proxy_object, Mo
  * the instance of a callvirt of \p method.
  */
 MonoMethod*
-mono_object_get_virtual_method (MonoObject *obj_raw, MonoMethod *method)
+mono_object_get_virtual_method_internal (MonoObject *obj_raw, MonoMethod *method)
 {
 	HANDLE_FUNCTION_ENTER ();
 	MonoMethod *result;
-	MONO_ENTER_GC_UNSAFE;
```
Comment (jaykrell) [inline mono/metadata/object.c]: Perhaps this PR should not have any GC mode change changes, i.e. do them in a separate PR.

File: mono/metadata/object.c [hunk 6]
```diff
@@ -2834,19 +2852,23 @@ mono_upgrade_remote_class (MonoDomain *domain, MonoObjectHandle proxy_object, Mo
...
 	ERROR_DECL (error);
 	MONO_HANDLE_DCL (MonoObject, obj);
 	result = mono_object_handle_get_virtual_method (obj, method, error);
 	mono_error_assert_ok (error);
-	MONO_EXIT_GC_UNSAFE;
```
Comment (jaykrell) [inline mono/metadata/object.c]: Perhaps this PR should not have any GC mode change changes, i.e. do them in a separate PR.

File: mono/metadata/object.c [hunk 7]
```diff
@@ -3309,19 +3331,21 @@ mono_copy_value (MonoType *type, void *dest, void *value, int deref_pointer)
  * The value must be in the native format of the field type. 
  */
 void
-mono_field_set_value (MonoObject *obj, MonoClassField *field, void *value)
+mono_field_set_value_internal (MonoObject *obj, MonoClassField *field, void *value)
 {
-	MONO_ENTER_GC_UNSAFE;
```
Comment (jaykrell) [inline mono/metadata/object.c]: Perhaps this PR should not have any GC mode change changes, i.e. do them in a separate PR.

File: mono/metadata/object.c [hunk 8]
```diff
@@ -3333,17 +3357,15 @@ mono_field_set_value (MonoObject *obj, MonoClassField *field, void *value)
  * The value must be in the native format of the field type. 
  */
 void
-mono_field_static_set_value (MonoVTable *vt, MonoClassField *field, void *value)
+mono_field_static_set_value_internal (MonoVTable *vt, MonoClassField *field, void *value)
 {
-	MONO_ENTER_GC_UNSAFE;
```
Comment (jaykrell) [inline mono/metadata/object.c]: Perhaps this PR should not have any GC mode change changes, i.e. do them in a separate PR.

File: mono/metadata/object.c [hunk 9]
```diff
@@ -3421,15 +3447,13 @@ mono_field_get_addr (MonoObject *obj, MonoVTable *vt, MonoClassField *field)
...
  * int i;
  *
- * mono_field_get_value (obj, int_field, &i);
+ * mono_field_get_value_internal (obj, int_field, &i);
  * </pre>
  */
 void
 mono_field_get_value (MonoObject *obj, MonoClassField *field, void *value)
 {
-	MONO_ENTER_GC_UNSAFE;
```
Comment (jaykrell) [inline mono/metadata/object.c]: Perhaps this PR should not have any GC mode change changes, i.e. do them in a separate PR.

**Comment by jaykrell** [comment]
@monojenkins build Linux x64

**Comment by jaykrell** [comment]
@monojenkins build Linux x64

**Comment by jaykrell** [comment]
CI looking promising now.

File: mono/metadata/object.c [hunk 10]
```diff
@@ -54,6 +54,13 @@
 #include "cominterop.h"
 #include <mono/utils/w32api.h>
 #include <mono/utils/unlocked.h>
+#include "external-only.h"
+#include "monitor.h"
+
+// Exports do not work portably/reliably if compiled separately.
+// They must be in .o/.obj files otherwise referenced.
+// .def files work better in this regard.
+#include "external-only.c"
```
[the remark this answers is not in the document]
Comment (lambdageek) [reply mono/metadata/object.c]: > [Exports] must be in .o/.obj files otherwise referenced.

I have no idea what this means.

Comment (lambdageek) [inline mono/metadata/object.c]: Discussed with Jay.  `#include "external-only.c"` can go away.  We don't quite have the problem like in https://github.com/mono/mono/pull/11188/files#r226004001 because we don't build intermediate `.lib` on Windows.  We should revisit using `.def` files another day.

**Comment by jaykrell** [comment]
@zoltan since you raised the issue?

`#define` vs. edit?
`m_foo`?
`mi_foo`?

**Comment by jaykrell** [comment]
@monojenkins build  Linux i386

**Comment by jaykrell** [comment]
The Unix linker is behaving how I described the Windows linker works.
The functions in external-only.c are not found by the test, unless the include is restored.

**Comment by lambdageek** [comment]
> The Unix linker is behaving how I described the Windows linker works.
> The functions in external-only.c are not found by the test, unless the include is restored.

But object.c is in the same libraries as external-only.c  (isn't it?) how can that possibly change anything?

**Comment by jaykrell** [comment]
Because we happen to call stuff in object.c.

**Comment by jaykrell** [comment]
I'm *hoping* `#include "external-only.c"` is acceptable to reviewers but I am reluctant to *assume* it.

**Comment by jaykrell** [comment]
Is the object.c factor clear?

The linker does not crack open .a files eagerly, and read of all of their constituent .o files, to export everything proactively.

It only reads the .o files within .a files if they resolve unresolved symbols.

Listing extra .a files on the linker command line, or putting extra .o files in .a files, does not cause stuff to be exported.

Putting extra stuff in object.c does make a difference, because we already require symbols from object.c or other reason. The linker has to resolve the entry points (main, DllMain, and .def file) and references from .o files given directly on the command line (not in .a files), and from there there is a graph walk.

It might be different for executable vs. shared libraries.
It isn't for Windows linker -- Windows linker has a simpler adequate more consistent model,  executables and shared libraries are indistinguishable.

Or there is an "entire archive" functionality that kicks in sometimes.
I thought it was always on Unix but clearly not.
I can build a small example on Mac like the above Windows and it behaves like Windows, and this test using dlsym against external_only symbols demonstrates it.

**Comment by lambdageek** [comment]
Thanks for clarifying, Jay. Frustrating that we have to play these games with the linker, but it sounds like we don't have any options that are less involved. I don't mind the include now.

**Comment by jaykrell** [comment]
There was MONO_EXTERNAL_ONLY_GC_UNSAFE.
It served to somewhat automate the tedium of ENTER_GC_UNSAFE / EXIT_GC_UNSAFE.
Then, we don't always switch the modes, so there is MONO_EXTERNAL_ONLY w/o the GC_UNSAFE.

This serves to be easily move back and forth to have the GC mode change or not, and it serves to mark the function as external_only.

These are minor things and they can be changed.
All of the macros could go away.

The reason for VOID vs. regular is because, without additional known machinery,  the macro has to know to introduce a local to assign to and return.

As to external-only.c, again, if the runtime does not reference a symbol within a source/.o file, then exports within that source/.o file are not exported. Unless we use .def files. Or move the .o files out of the .a files and link to them directly.  Or move the symbols to files that do have used symbols. include kinda does that last step, while maintaining a source organization that isolates external_only symbols -- if that is even desirable.

If we don't wish to separate them, then they can go to gc.c or such.

These particular symbols, if placed near what they replace, would be duplicated approx three times each -- null-gc.c, boehm-gc.c sgen-gc.c.

We needed it before, or not, *if* there were source files that no used symbols and only external_only symbols, or if we didn't notice that external_only symbols were disappearing.
One of those conditions was probably always true -- by leaving the functions where they were, they'd be in source files with used symbols and therefore still get exported.

It is quite easy to miss the loss of external_only symbols in  CI in fact.
We only notice them here, because there is one test that uses dlopen/dlsym to get to them, rather than calling them more directly. If you call them directly, you get a compiler error and you move to approved symbols.

Perhaps the mistake is in creating the external-only.c file, which doesn't work well, but seemed like a good idea.

As well, there is the pattern of replacing mono_foo with mono_foo_internal.
There are several options here and I am open to several.
We could m_foo, mi_foo, macros, etc.

**Comment by jaykrell** [comment]
Anything else needed here?

Should I just move the external-only.c content to gc.c?
  Or continue in that direction in a later PR and move stuff to external-only.c?
  It had been said in the past that it might be good to move external_only code out somehow, so I figured I'd go ahead with that. Initially more was moved here but I dialed back due to 1. harder to read diff 2. test failures.

Remove some/all the macros?

**Comment by jaykrell** [comment]
@monojenkins build OS X i386

**Comment by jaykrell** [comment]
Restated: We did not previously have any entirely unused source files. External-only.c is a new type of file that way.

**Comment by jaykrell** [comment]
ok?

**Comment by jaykrell** [comment]
Speaking to @kumpera, he deferred to @lambdageek.
Given no changes since @lambdageek approval, seems ok then.

File: mono/metadata/external-only.c [hunk 11]
```diff
@@ -6,18 +6,100 @@
...
+ * prevent the object from being moved (this should be avoided as much as possible
+ * and this should be used only for shorts periods of time or performance will suffer).
+ * To create a weakref use mono_gchandle_new_weakref (): track_resurrection should
+ * usually be false (see the GC docs for more details).
+ * mono_gchandle_get_target () can be used to get the object referenced by both kinds
+ * of handle: for a weakref handle, if an object has been collected, it will return NULL.
  */
-uintptr_t
-mono_array_length (MonoArray *array)
+uint32_t
```
Comment (lambdageek) [inline mono/metadata/external-only.c]: Hm.  you should probably copy the API doc comments from `sgen-mono.c`

File: mono/metadata/object.c [hunk 12]
```diff
@@ -54,6 +54,13 @@
 #include "cominterop.h"
 #include <mono/utils/w32api.h>
 #include <mono/utils/unlocked.h>
+#include "external-only.h"
+#include "monitor.h"
+
+// Exports do not work portably/reliably if compiled separately.
+// They must be in .o/.obj files otherwise referenced.
+// .def files work better in this regard.
```
Comment (lambdageek) [inline mono/metadata/object.c]: Maybe it doesn't hurt to be a bit more verbose here.  Something like:

```
// If an object file full of exported functions that are not called anywhere is included in
// an intermediate static library, and the library is in turn linked into another binary or library, the
// external only object fill will be dropped.  As a workaround, we include "external-only.c" here.  Another
// workaround on Windows would be to use .def files to specify exports.  On Unix, we would need to link
// external-only.o directly into the final binary or shared library.
```

Comment (jaykrell) [reply mono/metadata/object.c]: Unix has .def files or equivalent also.

Comment (jaykrell) [reply mono/metadata/object.c]: And listing .o files on link command line works on Windows. I'll wordsmith a bit.

File: mono/metadata/external-only.c [hunk 13]
```diff
@@ -6,18 +6,140 @@
...
+mono_gchandle_new_weakref (MonoObject *obj, mono_bool track_resurrection)
+{
+	MONO_EXTERNAL_ONLY (uint32_t, mono_gchandle_new_weakref_internal (obj, track_resurrection));
+}
+
+/**
+ * mono_gchandle_get_target:
+ * \param gchandle a GCHandle's handle.
+ *
+ * The handle was previously created by calling \c mono_gchandle_new_internal or
```
Comment (lambdageek) [inline mono/metadata/external-only.c]:
```suggestion
 * The handle was previously created by calling \c mono_gchandle_new or
```

File: mono/metadata/domain.c [hunk 14]
```diff
@@ -271,6 +278,12 @@ mono_string_hash (MonoString *s)
 	return h;	
 }
 
+guint
+mono_string_hash (MonoString *s)
```
Comment (lambdageek) [inline mono/metadata/domain.c]: Lost api doc comment.

File: mono/metadata/domain.c [hunk 15]
```diff
@@ -236,17 +237,23 @@ mono_install_free_domain_hook (MonoFreeDomainFunc func)
...
 		return TRUE;
 	if (l1 != l2)
 		return FALSE;
 
-	return memcmp (mono_string_chars (s1), mono_string_chars (s2), l1 * 2) == 0; 
+	return memcmp (mono_string_chars_internal (s1), mono_string_chars_internal (s2), l1 * 2) == 0;
+}
+
+gboolean
+mono_string_equal (MonoString *s1, MonoString *s2)
```
Comment (lambdageek) [inline mono/metadata/domain.c]: lost API doc comment.

File: mono/metadata/handle.h [hunk 16]
```diff
@@ -614,24 +606,12 @@ mono_gchandle_from_handle (MonoObjectHandle handle, mono_bool pinned);
...
 
-static inline gboolean
-mono_gchandle_target_equal (uint32_t gchandle, MonoObjectHandle equal)
-{
-	// This function serves to reduce coop handle creation.
-	MONO_HANDLE_SUPPRESS_SCOPE (1);
-	return mono_gchandle_get_target (gchandle) == MONO_HANDLE_RAW (equal);
-}
+gboolean
+mono_gchandle_target_equal (uint32_t gchandle, MonoObjectHandle equal);
```
Comment (lambdageek) [inline mono/metadata/handle.h]: Why? this isn't a public header

Comment (lambdageek) [reply mono/metadata/handle.h]: You can just update this inline function (and the others in handle.h to call the _internal functions.

File: mono/metadata/monitor.c [hunk 17]
```diff
@@ -634,6 +635,12 @@ mono_object_hash (MonoObject* obj)
 #endif
 }
 
+int
+mono_object_hash (MonoObject* obj)
```
Comment (lambdageek) [inline mono/metadata/monitor.c]: api doc comment.

File: mono/metadata/object.c [hunk 18]
```diff
@@ -2834,19 +2856,23 @@ mono_upgrade_remote_class (MonoDomain *domain, MonoObjectHandle proxy_object, Mo
...
 	ERROR_DECL (error);
 	MONO_HANDLE_DCL (MonoObject, obj);
 	result = mono_object_handle_get_virtual_method (obj, method, error);
 	mono_error_assert_ok (error);
-	MONO_EXIT_GC_UNSAFE;
 	HANDLE_FUNCTION_RETURN_VAL (result);
 }
 
+MonoMethod*
+mono_object_get_virtual_method (MonoObject *obj, MonoMethod *method)
```
Comment (lambdageek) [inline mono/metadata/object.c]: api doc comment.

File: mono/metadata/object.c [hunk 19]
```diff
@@ -3309,19 +3335,21 @@ mono_copy_value (MonoType *type, void *dest, void *value, int deref_pointer)
...
+		return;
 
 	dest = (char*)obj + field->offset;
 	mono_copy_value (field->type, dest, value, FALSE);
-leave:
-	MONO_EXIT_GC_UNSAFE;
+}
+
+void
+mono_field_set_value (MonoObject *obj, MonoClassField *field, void *value)
```
Comment (lambdageek) [inline mono/metadata/object.c]: api doc comment.

File: mono/metadata/object.c [hunk 20]
```diff
@@ -3357,8 +3383,12 @@ mono_field_static_set_value (MonoVTable *vt, MonoClassField *field, void *value)
 		dest = (char*)mono_vtable_get_static_field_data (vt) + field->offset;
 	}
 	mono_copy_value (field->type, dest, value, FALSE);
-leave:
-	MONO_EXIT_GC_UNSAFE;
+}
+
+void
+mono_field_static_set_value (MonoVTable *vt, MonoClassField *field, void *value)
```
Comment (lambdageek) [inline mono/metadata/object.c]: api doc comment

File: mono/metadata/object.c [hunk 21]
```diff
@@ -3421,15 +3451,13 @@ mono_field_get_addr (MonoObject *obj, MonoVTable *vt, MonoClassField *field)
  * <pre>
  * int i;
  *
- * mono_field_get_value (obj, int_field, &i);
+ * mono_field_get_value_internal (obj, int_field, &i);
```
Comment (lambdageek) [inline mono/metadata/object.c]:
```suggestion
 * mono_field_get_value (obj, int_field, &i);
```

File: mono/metadata/object.c [hunk 22]
```diff
@@ -4067,6 +4103,12 @@ mono_get_delegate_invoke (MonoClass *klass)
 	return result;
 }
 
+MonoMethod*
+mono_get_delegate_invoke (MonoClass *klass)
```
Comment (lambdageek) [inline mono/metadata/object.c]: api doc comment

File: mono/metadata/object.c [hunk 23]
```diff
@@ -4106,6 +4148,12 @@ mono_get_delegate_begin_invoke (MonoClass *klass)
 	return result;
 }
 
+MonoMethod*
+mono_get_delegate_begin_invoke (MonoClass *klass)
```
Comment (lambdageek) [inline mono/metadata/object.c]: api doc comment

File: mono/metadata/object.c [hunk 24]
```diff
@@ -4145,6 +4193,12 @@ mono_get_delegate_end_invoke (MonoClass *klass)
 	return result;
 }
 
+MonoMethod*
+mono_get_delegate_end_invoke (MonoClass *klass)
```
Comment (lambdageek) [inline mono/metadata/object.c]: api doc comment

File: mono/metadata/object.c [hunk 25]
```diff
@@ -4796,6 +4850,12 @@ mono_unhandled_exception (MonoObject *exc_raw)
 	HANDLE_FUNCTION_RETURN ();
 }
 
+void
+mono_unhandled_exception (MonoObject *exc)
```
Comment (lambdageek) [inline mono/metadata/object.c]: api doc comment

File: mono/metadata/object.c [hunk 26]
```diff
@@ -6363,17 +6422,23 @@ mono_string_empty_wrapper (void)
...
-mono_string_empty (MonoDomain *domain)
+mono_string_empty_internal (MonoDomain *domain)
 {
 	g_assert (domain);
 	g_assert (domain->empty_string);
 	return domain->empty_string;
 }
 
+MonoString*
+mono_string_empty (MonoDomain *domain)
```
Comment (lambdageek) [inline mono/metadata/object.c]: api doc comment

File: mono/metadata/object.c [hunk 27]
```diff
@@ -6720,14 +6781,15 @@ mono_string_new_wtf8_len_checked (MonoDomain *domain, const char *text, guint le
...
-	MonoString *res;
-	MONO_ENTER_GC_UNSAFE;
+	return mono_string_new_internal (mono_domain_get (), text);
+}
 
-	res = mono_string_new_internal (mono_domain_get (), text);
-	MONO_EXIT_GC_UNSAFE;
-	return res;
+MonoString*
+mono_string_new_wrapper (const char *text)
```
Comment (lambdageek) [inline mono/metadata/object.c]: api doc comment

File: mono/metadata/object.c [hunk 28]
```diff
@@ -6890,11 +6952,17 @@ mono_value_box_checked (MonoDomain *domain, MonoClass *klass, gpointer value, Mo
...
+mono_value_copy_internal (gpointer dest, gpointer src, MonoClass *klass)
 {
 	MONO_REQ_GC_UNSAFE_MODE;
 
-	mono_gc_wbarrier_value_copy (dest, src, 1, klass);
+	mono_gc_wbarrier_value_copy_internal (dest, src, 1, klass);
+}
+
+void
+mono_value_copy (gpointer dest, gpointer src, MonoClass *klass)
```
Comment (lambdageek) [inline mono/metadata/object.c]: api doc comment

File: mono/metadata/object.c [hunk 29]
```diff
@@ -6908,36 +6976,54 @@ mono_value_copy (gpointer dest, gpointer src, MonoClass *klass)
...
 
 	int size = mono_array_element_size (dest->obj.vtable->klass);
 	char *d = mono_array_addr_with_size_fast (dest, size, dest_idx);
 	g_assert (size == mono_class_value_size (m_class_get_element_class (mono_object_class (dest)), NULL));
-	mono_gc_wbarrier_value_copy (d, src, count, m_class_get_element_class (mono_object_class (dest)));
+	mono_gc_wbarrier_value_copy_internal (d, src, count, m_class_get_element_class (mono_object_class (dest)));
+}
+
+void
+mono_value_copy_array (MonoArray *dest, int dest_idx, void* src, int count)
```
Comment (lambdageek) [inline mono/metadata/object.c]: api doc comment

File: mono/metadata/object.c [hunk 30]
```diff
@@ -6908,36 +6976,54 @@ mono_value_copy (gpointer dest, gpointer src, MonoClass *klass)
...
-mono_object_get_domain (MonoObject *obj)
+mono_object_get_domain_internal (MonoObject *obj)
 {
 	MONO_REQ_GC_UNSAFE_MODE;
 
 	return mono_object_domain (obj);
 }
 
+MonoDomain*
+mono_object_get_domain (MonoObject *obj)
```
Comment (lambdageek) [inline mono/metadata/object.c]: api doc comment

File: mono/metadata/object.c [hunk 31]
```diff
@@ -6981,6 +7063,12 @@ mono_object_get_size (MonoObject* o)
 	}
 }
 
+unsigned
+mono_object_get_size (MonoObject *o)
```
Comment (lambdageek) [inline mono/metadata/object.c]: api doc comment

File: mono/metadata/object.c [hunk 32]
```diff
@@ -6990,22 +7078,23 @@ mono_object_get_size (MonoObject* o)
...
 	/* add assert for valuetypes? */
 	g_assert (m_class_is_valuetype (mono_object_class (obj)));
-	res = mono_object_get_data (obj);
-	MONO_EXIT_GC_UNSAFE;
-	return res;
+	return mono_object_get_data (obj);
+}
+
+void*
+mono_object_unbox (MonoObject *obj)
```
Comment (lambdageek) [inline mono/metadata/object.c]: api doc comment

File: mono/metadata/object.c [hunk 33]
```diff
@@ -7275,6 +7363,12 @@ mono_string_is_interned (MonoString *o)
 	return result;
 }
 
+MonoString*
+mono_string_is_interned (MonoString *str)
```
Comment (lambdageek) [inline mono/metadata/object.c]: api doc comment

File: mono/metadata/object.c [hunk 34]
```diff
@@ -7555,13 +7647,19 @@ mono_string_to_utf8_checked (MonoString *s, MonoError *error)
 	if (!s->length)
 		return g_strdup ("");
 
-	return mono_utf16_to_utf8 (mono_string_chars (s), s->length, error);
+	return mono_utf16_to_utf8 (mono_string_chars_internal (s), s->length, error);
+}
+
+char*
+mono_string_to_utf8_checked (MonoString *string_obj, MonoError *error)
```
Comment (lambdageek) [inline mono/metadata/object.c]: api doc comment

File: mono/metadata/object.c [hunk 35]
```diff
@@ -7620,19 +7718,25 @@ mono_string_to_utf16 (MonoString *s)
...
 		as [length] = 0;
 		if (length)
-			memcpy (as, mono_string_chars(s), length * sizeof (*as));
+			memcpy (as, mono_string_chars_internal (s), length * sizeof (*as));
 	}
 	return as;
 }
 
+mono_unichar2*
+mono_string_to_utf16 (MonoString *string_obj)
```
Comment (lambdageek) [inline mono/metadata/object.c]: api doc comment

File: mono/metadata/object.c [hunk 36]
```diff
@@ -7642,6 +7746,12 @@ mono_string_to_utf32 (MonoString *s)
 	return g_utf16_to_ucs4 (s->chars, s->length, NULL, NULL, NULL);
 }
 
+mono_unichar4*
+mono_string_to_utf32 (MonoString *string_obj)
```
Comment (lambdageek) [inline mono/metadata/object.c]: api doc comment

File: mono/metadata/object.c [hunk 37]
```diff
@@ -7808,14 +7918,20 @@ mono_get_eh_callbacks (void)
...
+mono_raise_exception_internal (MonoException *ex) 
 {
 	/* raise_exception doesn't return, so the transition to GC Unsafe is unbalanced */
 	MONO_STACKDATA (stackdata);
 	mono_threads_enter_gc_unsafe_region_unbalanced_with_info (mono_thread_info_current (), &stackdata);
 	mono_raise_exception_deprecated (ex);
 }
 
+void
+mono_raise_exception (MonoException *ex)
```
Comment (lambdageek) [inline mono/metadata/object.c]: api doc comment (IMPORTANT!)

File: mono/metadata/object.c [hunk 38]
```diff
@@ -8342,6 +8458,12 @@ mono_print_unhandled_exception (MonoObject *exc)
 		g_free (message);
 }
 
+void
+mono_print_unhandled_exception (MonoObject *exc)
```
Comment (lambdageek) [inline mono/metadata/object.c]: api doc comment

**Comment by lambdageek** [comment]
I only commented on about 25 of the external-only functions in object.c but in general: if you add an _internal version of a function, keep the API doc comment together with the non-_internal version - that's the one that will be the API documentation.

File: mono/metadata/domain.c (hunk 15 above)
Comment (jaykrell) [reply mono/metadata/domain.c]: Agreed, will fix.

File: mono/metadata/handle.h (hunk 16 above)
Comment (jaykrell) [reply mono/metadata/handle.h]: Cannot include object-internals.h without causing more breaks, and didn't work them through.

File: mono/metadata/monitor.c (hunk 17 above)
Comment (jaykrell) [reply mono/metadata/monitor.c]: Agreed, will fix.a

File: mono/metadata/object.c (hunk 18 above)
Comment (jaykrell) [reply mono/metadata/object.c]: Agreed, will fix.

File: mono/metadata/object.c (hunk 34 above)
Comment (jaykrell) [reply mono/metadata/object.c]: Agreed, will fix.

File: mono/metadata/object.c (hunk 35 above)
Comment (jaykrell) [reply mono/metadata/object.c]: Agreed, will fix.

**Comment by jaykrell** [comment]
Will duplicate (or move?) all the doc comments.

File: mono/metadata/handle.h (hunk 16 above)
Comment (lambdageek) [reply mono/metadata/handle.h]: Ah ok.  gotcha.

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by jaykrell** [comment]
@monojenkins build  Linux WebAssembly
@monojenkins build  Linux OS X x64 C++

**Comment by jaykrell** [comment]
@monojenkins build Linux WebAssembly

**Comment by jaykrell** [comment]
@monojenkins build Linux OS X x64 C++

**Comment by jaykrell** [comment]
@monojenkins build OS X x64 C++
